🚀 1단계 - 리팩터링(상품)

mj950425 opened the pull request:
안녕하세요 리뷰어님~

이번 미션 잘부탁드립니다!

mj19950425 committed:
feature : step1

kyucumber commented on src/main/kotlin/kitchenpos/tobe/product/domain/vo/ProductName.kt:
```diff
@@ -0,0 +1,36 @@
+import kitchenpos.tobe.product.exception.ProductNameException
+
+@Embeddable
+data class ProductName(
+    @Column(name = "name", nullable = false)
+    val name: String,
+) {
+    companion object {
+        fun of(
+            name: String,
+            productPurgomalumClient: ProductPurgomalumClient,
+        ): ProductName {
```
정적 팩터리가 있지만, 생성자 또한 열려있어서 생성이 가능해보이네요.

그리고 data class는 생성자를 private으로 하더라도 copy 등을 수행할 수 있기 때문에 완벽하게 안전하게 사용할 순 없습니다. 이 부분을 생각하면 그냥 class로 생성하는것도 고려해볼 수 있을 것 같네요.

mj950425 replied:
data class의 copy 기능이 편리하므로, data class로 만들어야지 했었는데.. 

그런 단점들이 존재하군요. class로 수정해두겠습니다.

kyucumber commented on src/main/kotlin/kitchenpos/tobe/infra/DefaultProductPurgomalumClient.kt:
```diff
@@ -0,0 +1,21 @@
+package kitchenpos.tobe.infra
```
구현체는 infra, 인터페이스는 domain에 잘 위치시키신 것 같네요! 👍

mj950425 replied:
감사합니다!

kyucumber commented on src/main/kotlin/kitchenpos/tobe/product/application/ProductServiceV2.kt:
```diff
@@ -0,0 +1,61 @@
+import org.springframework.stereotype.Service
+import java.util.*
+
+@Service
```
Transactional 어노테이션이 추가되어야 할 것 같네요.

kyucumber commented on src/main/kotlin/kitchenpos/tobe/product/application/ProductServiceV2.kt:
```diff
@@ -0,0 +1,61 @@
+        request: ChangeProductPriceRequest,
+    ): ChangeProductPriceResponse {
+        val product =
+            productRepository.findById(productId).orElseThrow {
+                throw IllegalArgumentException("상품을 찾을 수 없습니다.")
+            }
```
코틀린은 Optional을 쓸 필요가 없습니다.

findByIdOrNull을 사용하고 Elvis operator(?:)를 사용하는건 어떨까요?

mj950425 replied:
수정해두겠습니다!

kyucumber commented on src/main/kotlin/kitchenpos/tobe/product/application/ProductServiceV2.kt:
```diff
@@ -0,0 +1,61 @@
+            productRepository.findById(productId).orElseThrow {
+                throw IllegalArgumentException("상품을 찾을 수 없습니다.")
+            }
+        val changed = product.changePrice(request.price)
```
> 상품의 가격이 변경될 때 메뉴의 가격이 메뉴에 속한 상품 금액의 합보다 크면 메뉴가 숨겨진다.

도메인 부분만 구현하는거라 꼭 하지 않아도 괜찮지만, 응용 서비스쪽 구현을 해주셔서 코멘트 남깁니다. 위의 요구사항이 구현되어야 할 것 같아요.

mj950425 replied:
놓친부분이 있군요. 반영해두겠습니다.

kyucumber commented on src/test/kotlin/kitchenpos/tobe/application/ProductServiceTest.kt:
```diff
@@ -0,0 +1,3 @@
+package kitchenpos.tobe.application
+
+class ProductServiceTest
```
불필요한 클래스는 제거해도 좋을 것 같아요!

kyucumber commented on src/test/kotlin/kitchenpos/tobe/domain/repository/FakeProductRepository.kt:
```diff
@@ -0,0 +1,26 @@
+import kitchenpos.tobe.product.domain.repository.ProductRepository
+import java.util.*
+
+class FakeProductRepository : ProductRepository {
```
Fake를 이용한 테스트 👍

kyucumber reviewed (changes requested):
안녕하세요

몇가지 코멘트 남겨두었으니 확인 부탁드려요.

mj19950425 committed:
feature : 리뷰 반영

- 상품의 가격 변경시에 이벤트 핸들링을 통한 조건에 따른 메뉴 비노출 기능 추가
- jpa repo Optional 제거
- 트랜잭션 어노테이션 추가
- vo data class를 class로 변경

kyucumber commented on src/main/kotlin/kitchenpos/tobe/menu/infra/MenuEventHandler.kt:
```diff
@@ -0,0 +1,22 @@
+class MenuEventHandler(
+    private val menuService: MenuServiceV2,
+) {
+    @Async
+    @TransactionalEventListener(
+        classes = [ProductPriceChangedEvent::class],
+        phase = TransactionPhase.AFTER_COMMIT,
+    )
+    fun handle(event: ProductPriceChangedEvent) {
+        menuService.handleProductPriceChanged(event)
```
이벤트를 통한 메뉴와 상품간 컨텍스트 분리 및 Async 사용 시 트랜잭션 처리를 잘 해주셨네요 👍

kyucumber reviewed (approved):
코멘트 남긴 상품 변경에 따른 메뉴쪽 처리를 잘 구현해주셨네요.

다음 단계 진행을 위해 머지할게요.
